worker/provisioner/provisioner_task.go

```diff
@@ -894,29 +904,21 @@ func (task *provisionerTask) startMachines(machines []*apiprovisioner.Machine) e
 			return err
 		}
 
-		go func(machine *apiprovisioner.Machine, dg []string, index int) {
+		go func(machine *apiprovisioner.Machine, dg []string) {
 			defer wg.Done()
 			pInfo, startInstanceParams, err := task.setupToStartMachine(machine)
 			if err != nil {
-				errMachines[index] = err
-				task.removeMachinesFromAZMap(machine)
 				return
 			}
 			if err := task.startMachine(machine, pInfo, startInstanceParams, dg); err != nil {
-				errMachines[index] = err
-				task.removeMachinesFromAZMap(machine)
+				task.removeMachineFromAZMap(machine)
```

Comment:
I think you had the right idea before, setting status and and removing from the map in one place. Sorry, I might have led you astray... too much to fit in my head at once.

Unless it makes testing more difficult, I'd suggest moving the setupToStartMachine call to inside startMachine, and then do this:

```go
if err := task.startMachine(machine, dg); err != nil {
    task.removeMachineFromAZMap(machine)
    task.setErrorStatus(machine, err)
}
```

then you can drop the removeMachineFromAZMap and setErrorStatus calls within the startMachine method, except any that would be called while looping back around through new AZs

Reply:
Some of the errors in startMachines() have been intended to bubble up and cause the provisioner task to restart and some haven't. Right now tests are passing and provisioning is working.  Perhaps this is a refinement that can be done after the initial work?

Reply:
OK, we can do this in a follow up.